cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkerProgramRunner.java

```diff
@@ -0,0 +1,78 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.distributed;
+
+import co.cask.cdap.api.worker.WorkerSpecification;
+import co.cask.cdap.app.ApplicationSpecification;
+import co.cask.cdap.app.program.Program;
+import co.cask.cdap.app.runtime.ProgramController;
+import co.cask.cdap.app.runtime.ProgramOptions;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.twill.AbortOnTimeoutEventHandler;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.base.Preconditions;
+import com.google.inject.Inject;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.twill.api.EventHandler;
+import org.apache.twill.api.TwillController;
+import org.apache.twill.api.TwillRunner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+
+/**
+ * Distributed ProgramRunner for Worker.
+ */
+public class DistributedWorkerProgramRunner extends AbstractDistributedProgramRunner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DistributedWorkerProgramRunner.class);
+
+  @Inject
+  DistributedWorkerProgramRunner(TwillRunner twillRunner, Configuration hConfig, CConfiguration cConfig) {
+    super(twillRunner, hConfig, cConfig);
+  }
+
+  @Override
+  protected ProgramController launch(Program program, ProgramOptions options, File hConfFile, File cConfFile,
+                                     ApplicationLauncher launcher) {
+    ApplicationSpecification appSpec = program.getApplicationSpecification();
+    Preconditions.checkNotNull(appSpec, "Missing application specification.");
+
+    ProgramType processorType = program.getType();
+    Preconditions.checkNotNull(processorType, "Missing processor type.");
+    Preconditions.checkArgument(processorType == ProgramType.WORKER, "Only WORKER process type is supported.");
+
+    final WorkerSpecification workerSpec = appSpec.getWorkers().get(program.getName());
+    Preconditions.checkNotNull(workerSpec, "Missing WorkerSpecification for %s", program.getName());
+
+    LOG.info("Launching distributed worker: {}:{}", program.getName(), workerSpec.getName());
+
+    TwillController controller = launcher.launch(new WorkerTwillApplication(program, workerSpec, hConfFile, cConfFile,
+                                                                            eventHandler));
+
+    DistributedWorkerInstanceUpdater instanceUpdater = new DistributedWorkerInstanceUpdater(program, controller);
+    return new WorkerTwillProgramController(program.getName(), controller, instanceUpdater).startListen();
```

Comment:
So here you can pass in `program` instead of just the name, and get rid of the `instanceUpdater`, right?